span.go

```diff
@@ -469,6 +469,8 @@ func setSamplingPriority(s *Span, value interface{}) bool {
 	if !ok {
 		return false
 	}
+	s.RLock()
+	defer s.RUnlock()
 	if val == 0 {
 		s.context.samplingState.unsetSampled()
 		s.context.samplingState.setFinal()
```

Comment:
In L493-495, I think the write lock is unnecessary, could be just `s.SpanContext().samplingState.setFirehose()`

I hate that Github doesn't allow commenting on arbitrary line.